### PR TITLE
feat: Support goog.scope transformations in gents.

### LINKED ISSUE #349: Support Closure files with `goog.scope`
I converted some .js files into TypeScript. I see that `goog.scope` is not removed in the generated file.

### PR DESCRIPTION
**Pull request by dpurp**
Summary:
Rewrite the AST, removing the `goog.scope` wrapper, so that the various module and class-based transformations can take place.

Partially Adresses #349

### CODE REVIEW

File: src/main/java/com/google/javascript/gents/CommentLinkingPass.java [hunk 1]
```diff
@@ -140,8 +143,9 @@ private void linkCommentToNode(Node n) {
     }
 
     /** Removes unneeded tags from the comment */
-    private String filterTags(String comment) {
-      for (Pattern p : COMMENT_REPLACEMENTS) {
+    private String filterTags(Type type, String comment) {
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/CommentLinkingPass.java]: rename the method & change its comment?

File: src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    // Extract the goog.scope contents, and add them to module being constructed.
+    Node blockOfScopeContents = n.getLastChild().getLastChild().getLastChild();
+    blockOfScopeContents.detachFromParent();
+
+    // Collect all of the children into a list before detaching them from the parent,
+    // since detaching breaks the "getNext()" chain.
+    List<Node> scopeChildren = new ArrayList<>();
+    @Nullable Node maybeChildToAdd = blockOfScopeContents.getFirstChild();
+    while (maybeChildToAdd != null) {
+      scopeChildren.add(maybeChildToAdd);
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java]: optional: I think you can collapse these two loops, by always taking the first child here (both in the while condition and the body), and then removing that child and appending it at the insertion point.

File: src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    // Now that we have the contents of the goog.scope, rewrite the AST.
+
+    // Create a marker so that we know where to insert the goog.scope contents.
+    Node insertAfterThisNode = n;
+
+    for (Node scopeChild : scopeChildren) {
+      scopeChild.detachFromParent();
+      n.getParent().addChildAfter(scopeChild, insertAfterThisNode);
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java]: Are comments going to survive this? See NodeComments. I think yes, but might be worth adding a test.

File: src/main/java/com/google/javascript/gents/TypeScriptGenerator.java [hunk 4]
```diff
@@ -146,6 +146,9 @@ void generateTypeScript() {
     Node externRoot = compiler.getRoot().getFirstChild();
     Node srcRoot = compiler.getRoot().getLastChild();
 
+    RemoveGoogScopePass amodulePrePass = new RemoveGoogScopePass(compiler);
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/TypeScriptGenerator.java]: amodule?

File: src/test/java/com/google/javascript/gents/singleTests/goog_scope.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+goog.provide("A.B");
+goog.provide("A.B.C");
+
+goog.scope(function() {
+
```
Comment (mprobst) [inline src/test/java/com/google/javascript/gents/singleTests/goog_scope.js]: a common pattern is to alias "require" results here (`var LocalName = goog.something.LocalName;`). Maybe add a test case?

File: src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.javascript.jscomp.NodeTraversal;
+import com.google.javascript.rhino.Node;
+import com.google.javascript.rhino.Token;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Compiler pass removing the {@code goog.scope} wrapper, so that the module and class
+ * transformations can take effect.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java]: Add a comment that this translation is legal, because we're turning all files into modules so `var`s in the `goog.scope` body remain local (not global)?

File: src/main/java/com/google/javascript/gents/CommentLinkingPass.java (hunk 1 above)
Comment (dpurp) [inline src/main/java/com/google/javascript/gents/CommentLinkingPass.java]: Done.

File: src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java (hunk 6 above)
Comment (dpurp) [inline src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java]: Good point.

Unfortunately, this breaks the "aggressive" exporting.  I'd like to leave this as a TODO for a future cl, if possible.

File: src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java (hunk 2 above)
Comment (dpurp) [inline src/main/java/com/google/javascript/gents/RemoveGoogScopePass.java]: Done. Thanks!

File: src/test/java/com/google/javascript/gents/singleTests/goog_scope.js (hunk 5 above)
Comment (dpurp) [inline src/test/java/com/google/javascript/gents/singleTests/goog_scope.js]: Good point.

Unfortunately, this breaks the "aggressive" exporting.  I'd like to leave this as a TODO for a future cl, if possible.

File: src/test/java/com/google/javascript/gents/singleTests/goog_scope.ts [hunk 7]
```diff
@@ -2,14 +2,14 @@ export class Foo {
...
-    return 4;
+  static foo(): string {
+    return 'this should create a static method on Foo, since it is NOT goog.provided';
   }
 
   // TODO: Aggressively export rather than create static methods/fields
-  static qux(): number {
-    return 4;
+  static qux(): string {
+    return 'this should be directly exported, rather than create a static method, ' +
```
Comment (mprobst) [inline src/test/java/com/google/javascript/gents/singleTests/goog_scope.ts]: It isn't though, is it? I don't mind that (I think exporting a static function of a class is a rare corner case), but maybe make it clearer language wise that this is not intended behaviour?

Comment (dpurp) [inline src/test/java/com/google/javascript/gents/singleTests/goog_scope.ts]: the TODO above was supposed to indicate this :)
